File: right_angle_stack_cpp/src/cone_mapper.cpp
// 锥桶建图节点：将局部感知锥桶变换到世界坐标系，合并邻近锥桶，发布全局地图

#include <cmath>
#include <rclcpp/rclcpp.hpp>
#include <fsd_common_msgs/msg/cone.hpp>
#include <fsd_common_msgs/msg/cone_detections.hpp>
#include <fsd_common_msgs/msg/map.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <visualization_msgs/msg/marker.hpp>
#include <visualization_msgs/msg/marker_array.hpp>
#include "right_angle_stack_cpp/utils.hpp"

using namespace right_angle_stack_cpp;

// 路标：世界坐标 + 被观测次数 + 置信度
struct Landmark
{
  double x, y;
  int count;
  double pose_confidence;
  double color_confidence;
};

class ConeMapper : public rclcpp::Node
{
public:
  ConeMapper()
  : Node("cone_mapper")
  {
    this->declare_parameter("perception_map_topic", "/perception/cones");
    this->declare_parameter("perception_detections_topic", "/perception/cone_detections");
    this->declare_parameter("merge_distance", 0.75);   // 锥桶合并距离阈值
    this->declare_parameter("world_frame", "world");
    this->declare_parameter("base_frame", "base_link");

    merge_distance_ = this->get_parameter("merge_distance").as_double();
    world_frame_ = this->get_parameter("world_frame").as_string();
    base_frame_ = this->get_parameter("base_frame").as_string();

    map_pub_ = this->create_publisher<fsd_common_msgs::msg::Map>("/estimation/slam/map", 10);
    marker_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>("/visualization/cone_map", 10);

    // 订阅定位位姿，用于坐标变换
    pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "/localization/pose", 10,
      std::bind(&ConeMapper::on_pose, this, std::placeholders::_1));

    // 订阅感知结果（Map 格式）
    map_sub_ = this->create_subscription<fsd_common_msgs::msg::Map>(
      this->get_parameter("perception_map_topic").as_string(), 10,
      std::bind(&ConeMapper::on_map_detection, this, std::placeholders::_1));

    // 订阅感知结果（ConeDetections 格式）
    det_sub_ = this->create_subscription<fsd_common_msgs::msg::ConeDetections>(
      this->get_parameter("perception_detections_topic").as_string(), 10,
      std::bind(&ConeMapper::on_cone_detections, this, std::placeholders::_1));

    // 定时发布地图和可视化
    timer_ = this->create_wall_timer(std::chrono::milliseconds(200),
      std::bind(&ConeMapper::publish_outputs, this));

    RCLCPP_INFO(this->get_logger(),
      "Cone mapper started. Local perception is transformed into the world ENU frame.");
  }

private:
  void on_pose(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    pose_x_ = msg->pose.position.x;
    pose_y_ = msg->pose.position.y;
    pose_yaw_ = quaternion_to_yaw(msg->pose.orientation);
    has_pose_ = true;
  }

  void on_map_detection(const fsd_common_msgs::msg::Map::SharedPtr msg)
  {
    process_cones(msg->header.frame_id, msg->cone_blue, "blue");
    process_cones(msg->header.frame_id, msg->cone_yellow, "yellow");
    process_cones(msg->header.frame_id, msg->cone_red, "red");
    process_cones(msg->header.frame_id, msg->cone_unknown, "unknown");
    publish_outputs();
  }

  void on_cone_detections(const fsd_common_msgs::msg::ConeDetections::SharedPtr msg)
  {
    process_cones(msg->header.frame_id, msg->cone_detections, "");
    publish_outputs();
  }

  // 处理一组锥桶：坐标变换 → 合并
  void process_cones(
    const std::string & frame_id,
    const std::vector<fsd_common_msgs::msg::Cone> & cones,
    const std::string & fallback_color)
  {
    for (const auto & cone : cones) {
      std::string color = cone.color.empty() ? fallback_color : cone.color;
      std::string key = color_key(color);
      double wx, wy;
      if (!to_world(frame_id, cone, wx, wy)) continue;
      merge_landmark(key, wx, wy, cone.pose_confidence, cone.color_confidence);
    }
  }

  // 将锥桶坐标转换到世界坐标系
  bool to_world(const std::string & frame_id, const fsd_common_msgs::msg::Cone & cone, double & wx, double & wy)
  {
    std::string frame = frame_id;
    if (!frame.empty() && frame[0] == '/') frame = frame.substr(1);
    if (frame == world_frame_ || frame == "map") {
      wx = cone.position.x;
      wy = cone.position.y;
      return true;
    }
    if (!has_pose_) return false;
    body_to_world(cone.position.x, cone.position.y, pose_x_, pose_y_, pose_yaw_, wx, wy);
    return true;
  }

  // 合并路标：最近邻匹配，距离小于阈值则加权平均更新，否则新增
  void merge_landmark(const std::string & color, double x, double y,
                      double pose_confidence, double color_confidence)
  {
    auto & bucket = landmarks_[color];
    Landmark * closest = nullptr;
    double best_distance = std::numeric_limits<double>::infinity();

    for (auto & lm : bucket) {
      double dist = std::hypot(lm.x - x, lm.y - y);
      if (dist < best_distance) {
        best_distance = dist;
        closest = &lm;
      }
    }

    if (closest == nullptr || best_distance > merge_distance_) {
      // 新增路标
      bucket.push_back({x, y, 1, pose_confidence, color_confidence});
      return;
    }

    // 加权平均更新已有路标，count 上限 30
    int count = std::min(closest->count + 1, 30);
    double alpha = 1.0 / count;
    closest->x = (1.0 - alpha) * closest->x + alpha * x;
    closest->y = (1.0 - alpha) * closest->y + alpha * y;
    closest->count = count;
    closest->pose_confidence = std::max(closest->pose_confidence, pose_confidence);
    closest->color_confidence = std::max(closest->color_confidence, color_confidence);
  }

  fsd_common_msgs::msg::Cone make_cone(const std::string & color, const Landmark & lm)
  {
    fsd_common_msgs::msg::Cone cone;
    cone.position.x = lm.x;
    cone.position.y = lm.y;
    cone.position.z = 0.0;
    cone.color = color;
    cone.pose_confidence = lm.pose_confidence;
    cone.color_confidence = lm.color_confidence;
    return cone;
  }

  // 发布全局地图 + RViz 可视化
  void publish_outputs()
  {
    auto stamp = this->now();
    fsd_common_msgs::msg::Map msg;
    msg.header.stamp = stamp;
    msg.header.frame_id = world_frame_;

    for (const auto & lm : landmarks_["blue"]) msg.cone_blue.push_back(make_cone("blue", lm));
    for (const auto & lm : landmarks_["yellow"]) msg.cone_yellow.push_back(make_cone("yellow", lm));
    for (const auto & lm : landmarks_["red"]) msg.cone_red.push_back(make_cone("red", lm));
    for (const auto & lm : landmarks_["unknown"]) msg.cone_unknown.push_back(make_cone("unknown", lm));
    map_pub_->publish(msg);

    // RViz Marker 可视化：每个锥桶画一个圆柱体
    visualization_msgs::msg::MarkerArray markers;
    visualization_msgs::msg::Marker clear;
    clear.header = msg.header;
    clear.action = visualization_msgs::msg::Marker::DELETEALL;
    markers.markers.push_back(clear);

    int marker_id = 0;
    for (const auto & [color, bucket] : landmarks_) {
      for (const auto & lm : bucket) {
        visualization_msgs::msg::Marker marker;
        marker.header = msg.header;
        marker.ns = color + "_cones";
        marker.id = marker_id++;
        marker.type = visualization_msgs::msg::Marker::CYLINDER;
        marker.action = visualization_msgs::msg::Marker::ADD;
        marker.pose.position.x = lm.x;
        marker.pose.position.y = lm.y;
        marker.pose.position.z = 0.28;
        marker.pose.orientation.w = 1.0;
        marker.scale.x = 0.35;
        marker.scale.y = 0.35;
        marker.scale.z = 0.56;
        set_marker_color(marker, color);
        markers.markers.push_back(marker);
      }
    }
    marker_pub_->publish(markers);
  }

  double merge_distance_;
  std::string world_frame_, base_frame_;
  double pose_x_ = 0.0, pose_y_ = 0.0, pose_yaw_ = 0.0;
  bool has_pose_ = false;
  std::map<std::string, std::vector<Landmark>> landmarks_;

  rclcpp::Publisher<fsd_common_msgs::msg::Map>::SharedPtr map_pub_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_pub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_;
  rclcpp::Subscription<fsd_common_msgs::msg::Map>::SharedPtr map_sub_;
  rclcpp::Subscription<fsd_common_msgs::msg::ConeDetections>::SharedPtr det_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ConeMapper>());
  rclcpp::shutdown();
  return 0;
}

File: right_angle_stack_mix/src/cone_mapper/cone_mapper.py
#!/usr/bin/env python3
import math

import rclpy
from fsd_common_msgs.msg import Cone, Map
from geometry_msgs.msg import PoseStamped
from rclpy.node import Node
from visualization_msgs.msg import Marker, MarkerArray


def quaternion_to_yaw(q):
    siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
    cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
    return math.atan2(siny_cosp, cosy_cosp)


def body_to_world(local_x, local_y, origin_x, origin_y, yaw):
    c = math.cos(yaw)
    s = math.sin(yaw)
    return (
        origin_x + c * local_x - s * local_y,
        origin_y + s * local_x + c * local_y,
    )


def set_marker_color(marker, color):
    marker.color.a = 0.95
    if color == 'blue':
        marker.color.r = 0.02
        marker.color.g = 0.16
        marker.color.b = 1.0
    elif color == 'yellow':
        marker.color.r = 1.0
        marker.color.g = 0.78
        marker.color.b = 0.02
    elif color == 'red':
        marker.color.r = 1.0
        marker.color.g = 0.04
        marker.color.b = 0.02
    else:
        marker.color.r = 0.8
        marker.color.g = 0.8
        marker.color.b = 0.8


class ConeMapper(Node):
    def __init__(self):
        super().__init__('cone_mapper')

        self.declare_parameter('perception_map_topic', '/perception/cones')
        self.declare_parameter('pose_topic', '/localization/pose')
        self.declare_parameter('merge_distance', 0.75)
        self.declare_parameter('marker_scale_x', 0.35)
        self.declare_parameter('marker_scale_y', 0.35)
        self.declare_parameter('marker_scale_z', 0.56)

        self.merge_distance = float(self.get_parameter('merge_distance').value)
        self.marker_scale_x = float(self.get_parameter('marker_scale_x').value)
        self.marker_scale_y = float(self.get_parameter('marker_scale_y').value)
        self.marker_scale_z = float(self.get_parameter('marker_scale_z').value)
        self.pose = None
        self.landmarks = {
            'blue': [],
            'yellow': [],
            'red': [],
            'unknown': [],
        }

        self.map_pub = self.create_publisher(Map, '/estimation/slam/map', 10)
        self.marker_pub = self.create_publisher(MarkerArray, '/visualization/cone_map', 10)

        self.create_subscription(
            PoseStamped,
            self.get_parameter('pose_topic').value,
            self.on_pose,
            10,
        )
        self.create_subscription(
            Map,
            self.get_parameter('perception_map_topic').value,
            self.on_map_detection,
            10,
        )
        self.create_timer(0.2, self.publish_outputs)
        self.get_logger().info('Cone mapper started. Local perception is transformed into the world ENU frame.')

    def on_pose(self, msg):
        self.pose = (
            msg.pose.position.x,
            msg.pose.position.y,
            quaternion_to_yaw(msg.pose.orientation),
        )

    def on_map_detection(self, msg):
        self.process_cones(msg.header.frame_id, msg.cone_red, 'red')
        self.process_cones(msg.header.frame_id, msg.cone_blue, 'blue')
        self.process_cones(msg.header.frame_id, msg.cone_yellow, 'yellow')
        self.process_cones(msg.header.frame_id, msg.cone_unknown, 'unknown')
        self.publish_outputs()

    def process_cones(self, frame_id, cones, fallback_color):
        for cone in cones:
            color = cone.color if cone.color else fallback_color
            transformed = self.to_world(frame_id, cone)
            if transformed is None:
                continue
            wx, wy = transformed
            self.merge_landmark(color, wx, wy, cone.pose_confidence, cone.color_confidence)

    def to_world(self, frame_id, cone):
        if frame_id == 'world':
            return cone.position.x, cone.position.y
        if self.pose is None:
            return None
        x, y, yaw = self.pose
        return body_to_world(cone.position.x, cone.position.y, x, y, yaw)

    def merge_landmark(self, color, x, y, pose_confidence, color_confidence):
        bucket = self.landmarks[color]
        closest = None
        best_distance = float('inf')
        for landmark in bucket:
            dist = math.hypot(landmark['x'] - x, landmark['y'] - y)
            if dist < best_distance:
                best_distance = dist
                closest = landmark

        if closest is None or best_distance > self.merge_distance:
            bucket.append({
                'x': x,
                'y': y,
                'count': 1,
                'pose_confidence': pose_confidence,
                'color_confidence': color_confidence,
            })
            return

        count = min(closest['count'] + 1, 30)
        alpha = 1.0 / count
        closest['x'] = (1.0 - alpha) * closest['x'] + alpha * x
        closest['y'] = (1.0 - alpha) * closest['y'] + alpha * y
        closest['count'] = count
        closest['pose_confidence'] = max(closest['pose_confidence'], pose_confidence)
        closest['color_confidence'] = max(closest['color_confidence'], color_confidence)

    def make_cone(self, color, landmark):
        cone = Cone()
        cone.position.x = landmark['x']
        cone.position.y = landmark['y']
        cone.position.z = 0.0
        cone.color = color
        cone.pose_confidence = float(landmark['pose_confidence'])
        cone.color_confidence = float(landmark['color_confidence'])
        return cone

    def publish_outputs(self):
        stamp = self.get_clock().now().to_msg()
        msg = Map()
        msg.header.stamp = stamp
        msg.header.frame_id = 'world'
        msg.cone_blue = [self.make_cone('blue', item) for item in self.landmarks['blue']]
        msg.cone_yellow = [self.make_cone('yellow', item) for item in self.landmarks['yellow']]
        msg.cone_red = [self.make_cone('red', item) for item in self.landmarks['red']]
        msg.cone_unknown = [self.make_cone('unknown', item) for item in self.landmarks['unknown']]
        self.map_pub.publish(msg)

        markers = MarkerArray()
        clear = Marker()
        clear.header = msg.header
        clear.action = Marker.DELETEALL
        markers.markers.append(clear)

        marker_id = 0
        for color, bucket in self.landmarks.items():
            for landmark in bucket:
                marker = Marker()
                marker.header = msg.header
                marker.ns = f'{color}_cones'
                marker.id = marker_id
                marker_id += 1
                marker.type = Marker.CYLINDER
                marker.action = Marker.ADD
                marker.pose.position.x = landmark['x']
                marker.pose.position.y = landmark['y']
                marker.pose.position.z = self.marker_scale_z / 2.0
                marker.pose.orientation.w = 1.0
                marker.scale.x = self.marker_scale_x
                marker.scale.y = self.marker_scale_y
                marker.scale.z = self.marker_scale_z
                set_marker_color(marker, color)
                markers.markers.append(marker)
        self.marker_pub.publish(markers)


def main(args=None):
    rclpy.init(args=args)
    node = ConeMapper()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()